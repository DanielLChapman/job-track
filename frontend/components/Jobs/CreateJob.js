import React from "react";
import PropTypes from "prop-types";
import useForm from "../../library/useForm";
import gql from "graphql-tag";
import { CURRENT_USER_QUERY } from "../User";
import { useMutation } from "@apollo/client";

const CREATE_JOB_MUTATION = gql`
    mutation CREATE_JOB_MUTATION(
        $name: String!
        $status: String
        $author: ID!
        $salaryExpectation: Int
        $notes: String
    ) {
        createJob(
            data: {
                name: $name
                status: $status
                author: { connect: { id: $author } }
                salaryExpectation: $salaryExpectation
                notes: $notes
            }
        ) {
            id
            name
        }
    }
`;

CreateJob.propTypes = {};

function CreateJob(props) {
    if (!props.user) {
        return <span>Please Log In</span>;
    }

    const { inputs, handleChange, clearForm, resetForm } = useForm({
        name: "",
        status: "waiting",
        author: props.user.id || "",
        salaryExpectation: 0,
        notes: "",
    });

    const [createJob, { data, error, loading }] = useMutation(
        CREATE_JOB_MUTATION,
        {
            variables: {
                ...inputs,
            },
            refetchQueries: [{ query: CURRENT_USER_QUERY }],
        }
    );

    const handleSubmit = async (e) => {
        e.preventDefault();
        const res = await createJob();
        clearForm();
        if (props.closeFunc) {
            props.closeFunc(false);
        }
    };

    return (
        <>
            {error && <span>Error, Please Try Again</span>}
            <form method="POST" onSubmit={handleSubmit}>
                <h2>Create a new Job</h2>
                <fieldset disabled={loading} aria-busy={loading}>
                    <div className="input">
                        <input
                            type="text"
                            name="name"
                            placeholder="Name"
                            value={inputs.name}
                            required
                            className="input_field"
                            onChange={handleChange}
                        />
                        <label htmlFor="name">Name</label>
                    </div>

                    <div className="input">
                        <select
                            name="status"
                            value={inputs.status}
                            onChange={handleChange}
                        >
                            <option value="waiting">Waiting</option>
                            <option value="rejected">Rejected</option>
                            <option value="accepted">Accepted</option>
                        </select>
                        <label htmlFor="status">Status</label>
                    </div>
                    <div className="input">
                        <input
                            type="number"
                            name="salaryExpectation"
                            placeholder="0"
                            value={inputs.salaryExpectation}
                            onChange={handleChange}
                        />
                        <label htmlFor="salaryExpectation">
                            Salary Expectation
                        </label>
                    </div>
                    <div className="input">
                        <textarea
                            value={inputs.notes}
                            onChange={handleChange}
                            name="notes"
                        />

                        <label htmlFor="notes">Notes</label>
                    </div>

                    <button type="submit">Add Job</button>
                </fieldset>
            </form>
        </>
    );
}

export default CreateJob;
