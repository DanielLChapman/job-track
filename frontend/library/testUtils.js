import casual from 'casual';

casual.seed(231313);

const fakeUser = (overrides) => ({
    __typename: 'User',
    id: '4234',
    name: casual.name,
    jobs: [],
    email: casual.email,
    ...overrides,
  });

  const fakeJob = (overrides) => ({
    __typename: 'User',
    id: '4234',
    name: casual.name,
    publishedAt: casual.date('YYYY-MM-DD'),
    applicationDate: casual.date('YYYY-MM-DD'),
    status: "waiting",
    author: fakeUser().id,
    salaryExpectation: 0,
    notes: casual.sentence,
    ...overrides,
  });

  export {
    fakeUser,
    fakeJob
  }
