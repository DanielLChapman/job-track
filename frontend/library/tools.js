export const convertNum = (num) => {
    let t = num.toString();
    if (t.length < 3) {
        t = "$" + t
        return t;
    }
    let result = [];
    let count = 1;
    for (let i = t.length - 1; i >= 0; i--) {
        result.unshift(t[i]);
        
        if (count % 3 === 0 && i !== 0) {
            result.unshift(',')
        }
        count++;
    }

    result.unshift('$')

    return result.join('');

}