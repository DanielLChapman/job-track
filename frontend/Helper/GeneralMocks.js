
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeUser } from '../library/testUtils';

const fakeUserObj = fakeUser();
const notSignedInMocks = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: { data: { authenticatedItem: null } },
    },
  ];
  

  const signedInMocks = [
    {
      request: { query: CURRENT_USER_QUERY },
      result: { data: { authenticatedItem: fakeUser() } },
    },
  ];

  export {
    notSignedInMocks,
    signedInMocks,
}