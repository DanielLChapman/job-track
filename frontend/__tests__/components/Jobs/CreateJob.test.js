import {act, screen, render} from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import {signedInMocks, notSignedInMocks} from '../../../Helper/GeneralMocks'
import wait from 'waait';

import { fakeJob, fakeUser } from '../../../library/testUtils';
import CreateJob, { CREATE_JOB_MUTATION } from '../../../components/Jobs/CreateJob';
import userEvent from '@testing-library/user-event';

const job = fakeJob();


const mocks = [
    ...signedInMocks,
    {
        request: {
            query: CREATE_JOB_MUTATION,
            variables: {
                name: job.name,
                status: job.status,
                author: job.author,
                salaryExpectation: job.salaryExpectation,
                notes: job.notes,
            },
        },
        result: {
            data: {
                createJob: {
                    id: job.id,
                    __typename: 'Job',
                    name: job.name
                }
            }
        }
    }
]

describe('it renders a job submit form and submits successfully', () => {
    it ('does not load unless signed in', async() => {
        const {container, debug} = render(
            <MockedProvider>
                <CreateJob />
            </MockedProvider>
        )

        await screen.findByText(
            'Please Log In'
        );

        expect(container).toMatchSnapshot();
    })

    it ('does load correctly unless signed in', async() => {
        const {container, debug} = render(
            <MockedProvider mocks={signedInMocks}>
                <CreateJob user={fakeUser()} />
            </MockedProvider>
        )

        await screen.findByText(
            'Create a new Job'
        );

        expect(container).toMatchSnapshot();
    })

    it('calls the mutation properly', async () => {
        const {container, debug} = render(
            <MockedProvider mocks={mocks}>
                <CreateJob user={fakeUser()} />
            </MockedProvider>
        )

        // Type into the boxes
        await userEvent.type(screen.getByPlaceholderText(/name/i), job.name);
        await userEvent.type(screen.getByTestId('notes'), job.notes);
      // Click the submit
        await userEvent.click(screen.getByText('Add Job'));
        await act(async () => {
          await wait(0);
        })
        
        const element = await screen.getByTestId('success-signup-message');
        expect(element);
        /*await screen.findByText(
          `Signed Up With`
        );*/
      });
})