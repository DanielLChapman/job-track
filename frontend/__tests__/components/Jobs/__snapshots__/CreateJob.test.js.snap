// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a job submit form and submits successfully does load correctly unless signed in 1`] = `
<div>
  <form
    method="POST"
  >
    <h2>
      Create a new Job
    </h2>
    <fieldset
      aria-busy="false"
    >
      <div
        class="input"
      >
        <input
          class="input_field"
          name="name"
          placeholder="Name"
          required=""
          type="text"
          value=""
        />
        <label
          for="name"
        >
          Name
        </label>
      </div>
      <div
        class="input"
      >
        <select
          name="status"
        >
          <option
            value="waiting"
          >
            Waiting
          </option>
          <option
            value="rejected"
          >
            Rejected
          </option>
          <option
            value="accepted"
          >
            Accepted
          </option>
        </select>
        <label
          for="status"
        >
          Status
        </label>
      </div>
      <div
        class="input"
      >
        <input
          data-testid="salaryExpectation"
          name="salaryExpectation"
          placeholder="0"
          type="number"
          value="0"
        />
        <label
          for="salaryExpectation"
        >
          Salary Expectation
        </label>
      </div>
      <div
        class="input"
      >
        <textarea
          alt="textarea"
          data-testid="notes"
          name="notes"
        />
        <label
          for="notes"
        >
          Notes
        </label>
      </div>
      <button
        type="submit"
      >
        Add Job
      </button>
    </fieldset>
  </form>
</div>
`;

exports[`it renders a job submit form and submits successfully does not load unless signed in 1`] = `
<div>
  <span>
    Please Log In
  </span>
</div>
`;
